### ISSUE TITLE
Missing some predicates in ENSURE due to inheritance

### ISSUE THREAD

**Issue by rakshitkr** [CROSSINGTUD/CryptoAnalysis#114]
The below test case passes even though `init()` call is missing. This is because the analyzer considers the crySL rule of `AESLightEngine` for the first statement. The ORDER clause in the rule states that `init()` and `processBlock` are optional.  Thus only the predicate `generatedEngine` is ensured and ignores the second predicate i.e. `encrypted[buf]`. 

The call to `processBlock` is associated with `BlockCipher.processBlock ` method and hence will not be considered for the State Machine of `AESLightEngine`. Therefore the end result is that it's in accepting state and has ensured predicates.

```
@Test
	public void testAESLightEngine2() throws Exception {
		BlockCipher engine = new AESLightEngine();
		//no init
		byte[] buf = new byte[128];
		engine.processBlock(buf, 0, buf, 0);
		Assertions.hasEnsuredPredicate(engine);
		Assertions.mustBeInAcceptingState(engine);
	}
```

**Comment by johspaeth** [comment]
Is it also related to inheritance (#110)? Does the test pass, when we change 
`BlockCipher engine = new AESLightEngine();`
to
`AESLightEngine engine = new AESLightEngine();`
?

**Comment by rakshitkr** [comment]
Yes @johspaeth

**Comment by johspaeth** [comment]
Explanation:

In the case (1)
`AESLightEngine engine = new AESLightEngine();`, the call to `processBlock` resolves in Jimple to
` virtualinvoke engine.<org.bouncycastle.crypto.engines.AESLightEngine: int processBlock(byte[],int,byte[],int)>(buf, varReplacer45, buf, varReplacer45);`
Opposed to in (2): 
`BlockCipher engine = new AESLightEngine()` the call statically is mapped to ` interfaceinvoke engine.<org.bouncycastle.crypto.BlockCipher: int processBlock(byte[],int,byte[],int)>(buf, varReplacer45, buf, varReplacer45);` in Jimple.

The analysis currently does not match the `processBlock` call of the interface `BlockCipher`.
`

**Comment by johspaeth** [comment]
This is closed along with #110.

@rakshitkr that shouldn't block you anymore.
